README.md
<a href="http://www.boost.org/LICENSE_1_0.txt" target="_blank">![Boost Licence](http://img.shields.io/badge/license-boost-blue.svg)</a>
<a href="https://github.com/krzysztof-jusiak/di/releases" target="_blank">![Version](https://badge.fury.io/gh/krzysztof-jusiak%2Fdi.svg)</a>
<a href="https://github.com/krzysztof-jusiak/di/releases/latest" target="_blank">![Github Release](http://img.shields.io/github/release/krzysztof-jusiak/di.svg)</a>
<a href="http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/try_it_online.html" target="_blank">![Try Boost.DI Online](https://img.shields.io/badge/try%20boost.di-online-blue.svg)</a>
<a href="https://gitter.im/krzysztof-jusiak/di" target="_blank">![Gitter Chat](https://img.shields.io/badge/gitter-join%20chat%20%E2%86%92-brightgreen.svg)</a>

Boost.DI: C++ Dependency Injection
===============================================
[![Boost Libraries](https://raw.githubusercontent.com/krzysztof-jusiak/di/gh-pages/boost/boost.png)](http://www.boost.org)
<a href="https://travis-ci.org/krzysztof-jusiak/di" target="_blank">![Build Status](https://img.shields.io/travis/krzysztof-jusiak/di/cpp14.svg?label=linux/osx)</a>
<a href="https://ci.appveyor.com/project/krzysztof-jusiak/di" target="_blank">![Build Status](https://img.shields.io/appveyor/ci/krzysztof-jusiak/di/cpp14.svg?label=windows)</a>
<a href="https://coveralls.io/r/krzysztof-jusiak/di?branch=cpp14" target="_blank">![Coveralls](http://img.shields.io/coveralls/krzysztof-jusiak/di/cpp14.svg)</a>
<a href="http://github.com/krzysztof-jusiak/di/issues" target="_blank">![Github Issues](https://img.shields.io/github/issues/krzysztof-jusiak/di.svg)</a>

> "Don't call us, we'll call you", Hollywood principle

**[Dependency Injection](http://www.youtube.com/watch?v=IKD2-MAkXyQ) (DI)** involves passing (injecting) one or more dependencies (or services) to a dependent object (or client) which become part of the client’s state.
It is like the Strategy Pattern, except the strategy is set once, at construction. DI enables loosely coupled designs, which are easier to maintain and test.

> **"Let's make some coffee!"**
<p align="center"><img src="https://raw.githubusercontent.com/krzysztof-jusiak/di/cpp14/doc/images/coffee_maker.png" alt="coffee maker"/></p>
```cpp
No Dependency injection                 | Dependency Injection
----------------------------------------|--------------------------------------------
class coffee_maker {                    | class coffee_maker {
public:                                 | public:
    void brew() {                       |   coffee_maker(shared_ptr<iheater> heater
        heater->on();                   |              , unique_ptr<ipump> pump)
        pump->pump();                   |         : heater(heater), pump(move(pump))
        clog << "coffee"! << endl;      |     { }
        heater->off();                  |
    }                                   |     void brew() {
                                        |         heater->on();
private:                                |         pump->pump();
    shared_ptr<iheater> heater =        |         clog << "coffee!" << endl;
        make_shared<electric_heater>(); |         heater->off();
                                        |     }
    unique_ptr<ipump> pump =            |
        make_unique<heat_pump>(heater); | private:
};                                      |     shared_ptr<iheater> heater;
                                        |     unique_ptr<ipump> pump;
                                        | };
                                        |
```

**Why Dependency Injection?**

* DI provides loosely coupled code (separation of business logic and object creation)
* DI provides easier to maintain code (different objects might be easily injected)
* DI provides easier to test code (fakes objects might be injected)

[![The Clean Code Talks - Don't Look For Things!](http://img.youtube.com/vi/RlfLCWKxHJ0/0.jpg)](http://www.youtube.com/watch?v=RlfLCWKxHJ0) | [![DAGGER 2 - A New Type of dependency injection](http://image.slidesharecdn.com/nr73mexetqiybd1qpcem-140529143342-phpapp01/95/dependency-injection-for-android-5-638.jpg?cb=1401392113)](http://www.youtube.com/watch?v=oK_XtfXPkqw) |
--- | --- |

**Why Boost.DI?**

* Boost.DI has none or minimal run-time overhead - [Run-time performance](#run_time_performance)
* Boost.DI compiles fast - [Compile-time performance](#compile_time_performance)
* Boost.DI gives short diagnostic messages - [Diagnostic messages](#diagnostic_messages)
* Boost.DI is not intrusive [link to tutorial]
* Boost.DI reduces boilerplate code [link to tutorial]
* Boost.DI reduces testing effort [link to tutorial]
* Boost.DI gives better control of what and how is created [link to tutorial]
* Boost.DI gives better understanding about objects hierarchy [link to tutorial]

**How To Start?**

* Get C++14 compliant compiler (Clang-3.4+, GCC-5.1+, Visual Studio 2015+)
	* STL is not required (when available supports std::unique\_ptr, std::shared\_ptr, std::weak\_ptr)
	* Boost is not required (when available supports boost::shared\_ptr)
* [Read Quick User Guide](#quick_user_guide)
* [Read Tutorial](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/tutorial.html)
* [Read Documentation](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html)

> To get started the only file you need is `di.hpp`:

[https://raw.githubusercontent.com/krzysztof-jusiak/di/cpp14/include/boost/di.hpp](https://raw.githubusercontent.com/krzysztof-jusiak/di/cpp14/include/boost/di.hpp)

```sh
curl https://raw.githubusercontent.com/krzysztof-jusiak/di/cpp14/include/boost/di.hpp --create-dirs -o boost/di.hpp
```

```sh
cat << EOF > main.cpp
#include <boost/di.hpp>
int main() { }
EOF
```

```sh
$CXX -std=c++14 -I. -fno-exceptions -fno-rtti -nostdinc -nostdinc++ -pedantic -pedantic-errors -Wall -Wextra -Werror main.cpp
```

> To get and test Boost.DI library:

```sh
    git clone https://github.com/krzysztof-jusiak/di.git
    cd build && cmake ..
    make all test
```

<a id="quick_user_guide"></a>
**Quick User Guide** | [Examples](https://github.com/krzysztof-jusiak/di/tree/cpp14/example/quick_user_guide)

* [Injector](#injector) | [Bindings](#bindings) | [Injections](#injections) | [Annotations](#annotations) | [Scopes](#scopes) | [Modules](#modules) | [Providers](#providers) | [Policies](#policies)
* [Run-time performance](#run_time_performance) | [Compile-time performance](#compile_time_performance)
* [Diagnostic messages](#diagnostic_messages) | [Configuration](#configuration)

> Let's assume all examples below include `boost/di.hpp` header and define `di` namespace alias.
```cpp
#include <boost/di.hpp>
namespace di = boost::di;
//
struct i1 { virtual ~i1() = default; virtual void dummy1() = 0; };
struct i2 { virtual ~i2() = default; virtual void dummy2() = 0; };
struct impl1 : i1 { void dummy1() override { } };
struct impl2 : i2 { void dummy2() override { } };
struct impl : i1, i2 { void dummy1() override { } void dummy2() override { } };
```

*
<a id="injector"></a>
> **[Injector](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/injector.html)**
```cpp
Create empty injector                   | Test
----------------------------------------|-----------------------------------------
auto injector = di::make_injector();    | assert(0 == injector.create<int>());
```

*

<a id="bindings"></a>
> **[Bindings](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/bindings.html)** | [Examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/bindings.cpp) | [More examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/dynamic_bindings.cpp)
```cpp
Bind interface to implementation        | Test
----------------------------------------|-----------------------------------------
auto injector = di::make_injector(      | auto object = injector.create<unique_ptr<i1>>();
    di::bind<i1>.to<impl1>()            | assert(dynamic_cast<impl1*>(object.get()));
);                                      |
```
```cpp
Bind different interfaces to one        | Test
implementation                          |
----------------------------------------|-----------------------------------------
auto injector = di::make_injector(      | auto object1 = injector.create<shared_ptr<i1>>();
    di::bind<i1, i2>.to<impl>()         | auto object2 = injector.create<shared_ptr<i2>>();
);                                      | assert(dynamic_cast<impl*>(object1.get()));
                                        | assert(dynamic_cast<impl*>(object2.get()));
```
```cpp
Bind type to value                      | Test
----------------------------------------|-----------------------------------------
auto injector = di::make_injector(      | assert(42 == injector.create<int>());
    di::bind<int>.to(42)                |
);                                      |
```
```cpp
Cross platform `bind` (Clang/GCC/MSVC)  | Test
----------------------------------------|-----------------------------------------
auto injector = di::make_injector(      | auto object = injector.create<unique_ptr<i1>>();
    di::bind<i1>().to<impl>()           | assert(dynamic_cast<impl1*>(object.get()));
    // MSVC 2015 doesnt support variable|
    // templates                        |
);                                      |
```
```cpp
Advanced binding                        | Test
----------------------------------------|-----------------------------------------
auto l = 42l;                           | assert(42 == injector.create<int>());
auto b = false;                         | assert(injector.create<shared_ptr<i1>>()
                                        |        ==
auto injector = di::make_injector(      |        injector.create<shared_ptr<i1>>()
   di::bind<int>.to(0)                  | );
 , di::bind<int>.to(42) [di::override]  | assert(l == injector.create<long&>());
 , di::bind<i1>.to(make_shared<impl>()) | assert(&l == &injector.create<long&>());
 , di::bind<long>.to(l)                 | assert(87 == injector.create<short>());
 , di::bind<short>.to([]{return 87;})   | {
 , di::bind<i2>.to(                     | auto object = injector.create<shared_ptr<i2>>();
     [&](const auto& injector)          | assert(nullptr == object);
        -> shared_ptr<i2> {             | }
            if (b) {                    | {
              return injector.template  | b = true;
                create<                 | auto object = injector.create<shared_ptr<i2>>();
                  shared_ptr<impl2>>(); | assert(dynamic_cast<impl2*>(object.get()));
            }                           | }
            return nullptr;             |
     }                                  |
   )                                    |
);                                      |
```

*

<a id="injections"></a>
> **[Injections](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/injections.html)** | [Examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/constructor_injection.cpp) | [More examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/constructor_signature.cpp)
```cpp
Direct constructor injection            | Test
----------------------------------------|-----------------------------------------
struct c {                              | auto object = injector.create<c>();
    c(int a, double d) : a(a), d(d) { } | assert(42 == object.a);
                                        | assert(87.0 == object.d);
    int a = 0;                          |
    double d = 0.0;                     |
};                                      |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.to(42)                |
  , di::bind<double>.to(87.0)           |
);                                      |
```
```cpp
Aggregate constructor injection         | Test
----------------------------------------|-----------------------------------------
struct c {                              | auto object = injector.create<c>();
    int a = 0;                          | assert(42 == object.a);
    double d = 0.0;                     | assert(87.0 == object.d);
};                                      |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.to(42)                |
  , di::bind<double>.to(87.0)           |
);                                      |
```
```cpp
Direct constructor injection with many  | Test
constructors (longest parameters list   |
constructor will be chosen)             |
----------------------------------------|-----------------------------------------
struct c {                              | auto object = injector.create<c>();
    c();                                | assert(42 == object.a);
    c(int a) : a(a) { }                 | assert(87.0 == object.d);
    c(int a, double d) : a(a), d(d) { } |
                                        |
    int a = 0;                          |
    double d = 0.0;                     |
};                                      |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.to(42)                |
  , di::bind<double>.to(87.0)           |
);                                      |
```
```cpp
Direct constructor injection with       | Test
ambiguous constructors (BOOST_DI_INJECT)|
----------------------------------------|-----------------------------------------
struct c {                              | auto object = injector.create<c>();
    c(double d, int a) : a(a), d(d) { } | assert(42 == object.a);
    BOOST_DI_INJECT(c, int a, double d) | assert(87.0 == object.d);
        : a(a), d(d) { }                |
                                        |
    int a = 0;                          |
    double d = 0.0;                     |
};                                      |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.to(42)                |
  , di::bind<double>.to(87.0)           |
);                                      |
```
```cpp
Direct constructor injection with       | Test
ambiguous constructors                  |
(BOOST_DI_INJECT_TRAITS)                |
----------------------------------------|-----------------------------------------
struct c {                              | auto object = injector.create<c>();
    BOOST_DI_INJECT_TRAITS(int, double);| assert(42 == object.a);
    c(double d, int a) : a(a), d(d) { } | assert(87.0 == object.d);
    c(int a, double d) : a(a), d(d) { } |
                                        |
    int a = 0;                          |
    double d = 0.0;                     |
};                                      |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.to(42)                |
  , di::bind<double>.to(87.0)           |
);                                      |
```
```cpp
Direct constructor injection with       | Test
ambiguous constructors                  |
(di::ctor_traits)                       |
----------------------------------------|-----------------------------------------
struct c {                              | auto object = injector.create<c>();
    c(double d, int a) : a(a), d(d) { } | assert(42 == object.a);
    c(int a, double d) : a(a), d(d) { } | assert(87.0 == object.d);
                                        |
    int a = 0;                          |
    double d = 0.0;                     |
};                                      |
                                        |
namespace boost { namespace di {        |
template<>                              |
struct ctor_traits<c> {                 |
    BOOST_DI_INJECT_TRAITS(int, double);|
};                                      |
}} // boost::di                         |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.to(42)                |
  , di::bind<double>.to(87.0)           |
);                                      |
```

*

<a id="annotations"></a>
> **[Annotations](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/annotations.html)** | [Examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/annotations.cpp)
```cpp
Annotated constructor injection         | Test
----------------------------------------|-----------------------------------------
auto int1 = []{};                       | auto object = injector.create<c>();
auto int2 = []{};                       | assert(42 == object.a);
                                        | assert(87 == object.b);
struct c {                              |
    BOOST_DI_INJECT(c                   |
        , (named = int1) int a          |
        , (named = int2) int b)         |
        : a(a), b(b)                    |
    { }                                 |
                                        |
    int a = 0;                          |
    int b = 0;                          |
};                                      |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.named(int1).to(42)    |
  , di::bind<int>.named(int2).to(87)    |
);                                      |
```
```cpp
Annotated constructor injection with    | Test
the same names for different parameters |
----------------------------------------|-----------------------------------------
auto n1 = []{};                         | auto object = injector.create<c>();
auto n2 = []{};                         | assert(42 == object.i1);
                                        | assert(42 == object.i2);
struct c {                              | assert(87 == object.i3);
  BOOST_DI_INJECT(c                     | assert(0 == object.i4);
      , (named = n1) int a              | assert("str" == c.s);
      , (named = n1) int b              |
      , (named = n2) int c              |
      , int d                           |
      , (named = n1) string s)          |
  : i1(i1), i2(i2), i3(i3), i4(i4), s(s)|
  { }                                   |
                                        |
  int i1 = 0;                           |
  int i2 = 0;                           |
  int i3 = 0;                           |
  int i4 = 0;                           |
  string s;                             |
};                                      |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.named(n1).to(42)      |
  , di::bind<int>.named(n2).to(87)      |
  , di::bind<string>.named(n1).to("str")|
);                                      |
```
```cpp
Annotated constructor injection with    | Test
separate constructor definition         |
----------------------------------------|-----------------------------------------
auto int1 = []{};                       | auto object = injector.create<c>();
auto int2 = []{};                       | assert(42 == object.a);
                                        | assert(87 == object.b);
struct c {                              |
    BOOST_DI_INJECT(c                   |
        , (named = int1) int a          |
        , (named = int2) int b);        |
                                        |
    int a = 0;                          |
    int b = 0;                          |
};                                      |
                                        |
c::c(int a, int b) : a(a), b(b) { }     |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.named(int1).to(42)    |
  , di::bind<int>.named(int2).to(87)    |
);                                      |
```
```cpp
Annotated constructor injection with    | Test
di::ctor_traits                         |
----------------------------------------|-----------------------------------------
auto int1 = []{};                       | auto object = injector.create<c>();
auto int2 = []{};                       | assert(42 == object.a);
                                        | assert(87 == object.b);
struct c {                              |
    c(int a, int b) : a(a), b(b) { }    |
                                        |
    int a = 0;                          |
    int b = 0;                          |
};                                      |
                                        |
namespace boost { namespace di {        |
template<>                              |
struct ctor_traits<c> {                 |
    BOOST_DI_INJECT_TRAITS(             |
        (named = int1) int              |
      , (named = int2) int);            |
};                                      |
}} // boost::di                         |
                                        |
auto injector = di::make_injector(      |
    di::bind<int>.named(int1).to(42)    |
  , di::bind<int>.named(int2).to(87)    |
);                                      |
```

*

<a id="scopes"></a>
> **[Scopes](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/scopes.html)** | [Examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/deduce_scope.cpp) | [More examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/custom_scope.cpp)
```cpp
Deduce scope (default)                  | Test
----------------------------------------|-----------------------------------------
struct c {                              | auto object1 = injector.create<unique_ptr<c>>();
    shared_ptr<i1> sp; /*singleton*/    | auto object2 = injector.create<unique_ptr<c>>();
    unique_ptr<i2> up; /*unique*/       | assert(object1->sp == object2->sp);
    int& i; /*singleton*/               | assert(object1->up != object2->up);
    double d; /*unique*/                | assert(42 == object1->i);
};                                      | assert(&i == &object1->i;
                                        | assert(42 == object2->i);
auto i = 42;                            | assert(&i == &object2->i);
                                        | assert(87.0 == object1->d);
auto injector = di::make_injector(      | assert(87.0 == object2->d);
    di::bind<i1>.to<impl1>()            |
  , di::bind<i2>.to<impl2>()            |
  , di::bind<int>.to(i)                 |
  , di::bind<double>.to(87.0)           |
);                                      |
```
| Type | Deduced scope |
|------|-------|
| T | unique |
| T& | singleton |
| const T& | unique (temporary)/singleton |
| T* | unique (ownership transfer) |
| const T* | unique (ownership transfer) |
| T&& | unique |
| unique\_ptr<T> | unique |
| shared\_ptr<T> | singleton |
| weak\_ptr<T> | singleton |
```cpp
Unique scope                            | Test
----------------------------------------|-----------------------------------------
auto injector = di::make_injector(      | assert(injector.create<shared_ptr<i1>>()
 di::bind<i1>.in(di::unique).to<impl1>()|        !=
);                                      |        injector.create<shared_ptr<i1>>()
                                        | );
```
```cpp
Singleton scope (shared between threads)| Test
----------------------------------------|-----------------------------------------
auto injector = di::make_injector(      | assert(injector.create<shared_ptr<i1>>()
    di::bind<i1>.                       |        ==
        in(di::singleton).to<impl1>()   |        injector.create<shared_ptr<i1>>()
);                                      | );
```
```cpp
Custom scope                            | Test
----------------------------------------|-----------------------------------------
struct custom_scope {                   | assert(injector.create<shared_ptr<i1>>()
  template<class TExpected, class>      |        !=
  struct scope {                        |        injector.create<shared_ptr<i1>>()
    template<class T, class TProvider>  | );
    auto create(const TProvider& pr) {  |
      return                            |
        shared_ptr<TExpected>{pr.get()};|
    }                                   |
  };                                    |
};                                      |
                                        |
auto injector = di::make_injector(      |
    di::bind<i1>.                       |
        in(custom_scope{}).to<impl1>()  |
);                                      |
```
| Type/Scope | unique | singleton | external |
|------------|--------|--------|-----------|---------|----------|
| T | ✔ | - | ✔ |
| T& | - | ✔  | ✔ |
| const T& | ✔ (temporary) | ✔ | ✔ |
| T* (transfer ownership) | ✔ | - | - | - | ✔ |
| const T* | ✔ | - | ✔ |
| T&& | ✔ | - | - |
| unique\_ptr<T> | ✔ |  - | ✔ |
| shared\_ptr<T> | ✔ | ✔ | ✔ |
| weak\_ptr<T> | - | ✔ | ✔ |

*

<a id="modules"></a>
> **[Modules](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/modules.html)** | [Examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/modules.cpp)
```cpp
Module                                  | Test
----------------------------------------|-----------------------------------------
struct c {                              | auto object = injector.create<unique_ptr<c>>();
    c(unique_ptr<i1> i1                 | assert(dynamic_cast<impl1*>(object->i1.get()));
    , unique_ptr<i2> i2                 | assert(dynamic_cast<impl2*>(object->i2.get()));
    , int i) : i1(move(i1))             | assert(42 == object->i);
             , i2(move(i2)), i(i)       |
    { }                                 | auto up1 = injector.create<unique_ptr<i1>>();
                                        | assert(dynamic_cast<impl1*>(up1.get()));
    unique_ptr<i1> i1;                  |
    unique_ptr<i2> i2;                  | auto up2 = injector.create<unique_ptr<i2>>();
    int i = 0;                          | assert(dynamic_cast<impl2*>(up2.get()));
};                                      |
                                        |
struct module1 {                        |
    auto configure() const noexcept {   |
        return di::make_injector(       |
            di::bind<i1>.t<impl1>()     |
          , di::bind<int>.to(42)        |
        );                              |
    }                                   |
};                                      |
                                        |
struct module2 {                        |
    auto configure() const noexcept {   |
        return di::make_injector(       |
            di::bind<i2>.to<impl2>()    |
        );                              |
    };                                  |
};                                      |
                                        |
auto injector = di::make_injector(      |
    module1{}, module2{}                |
);                                      |
```
```cpp
Exposed type module                     | Test
----------------------------------------|-----------------------------------------
struct c {                              | auto object = injector.create<c>();
    c(shared_ptr<i1> i1                 | assert(dynamic_cast<impl1*>(object.i1.get()));
    , shared_ptr<i2> i2                 | assert(dynamic_cast<impl2*>(object.i2.get()));
    , int i) : i1(i1), i2(i2), i(i)     | assert(42 == object.i);
    { }                                 |
                                        | // injector.create<unique_ptr<i1>>() // compile error
    shared_ptr<i1> i1;                  | // injector.create<unique_ptr<i2>>() // compile error
    shared_ptr<i2> i2;                  |
    int i = 0;                          |
};                                      |
                                        |
struct module {                         |
    di::injector<c> configure()         |
    const noexcept;                     |
                                        |
    int i = 0;                          |
};                                      |
                                        |
di::injector<c> // expose c             |
module::configure() const noexcept {    |
    return di::make_injector(           |
        di::bind<i1>.to<impl1>()        |
      , di::bind<i2>.to<impl2>()        |
      , di::bind<int>.to(i)             |
    );                                  |
}                                       |
                                        |
auto injector = di::make_injector(      |
    module{42}                          |
);                                      |
```
```cpp
Exposed many types module               | Test
----------------------------------------|-----------------------------------------
struct module {                         | auto up1 = injector.create<unique_ptr<i1>>();
    di::injector<i1, i2> configure()    | assert(dynamic_cast<impl1*>(up1.get()));
    const noexcept;                     |
                                        | auto up2 = injector.create<unique_ptr<i2>>();
    int i = 0;                          | assert(dynamic_cast<impl2*>(up2.get()));
};                                      |
                                        |
di::injector<i1, i2> // expose i1, i2   |
module::configure() const noexcept {    |
    return di::make_injector(           |
        di::bind<i1>.to<impl1>()        |
      , di::bind<i2>.to<impl2>()        |
    );                                  |
}                                       |
                                        |
auto injector = di::make_injector(      |
    module{}                            |
);                                      |
```
```cpp
Exposed type module with annotation     | Test
----------------------------------------|-----------------------------------------
auto my = []{};                         | auto object = injector.create<unique_ptr<c>>();
                                        | assert(dynamic_cast<impl1*>(object->up.get()));
struct c {                              |
    BOOST_DI_INJECT(c                   |
      , (named = my) unique_ptr<i1> up) |
      : up(up)                          |
    { }                                 |
                                        |
    unique_ptr<i1> up;                  |
};                                      |
                                        |
di::injector<i1> module =               |
    di::make_injector(                  |
        di::bind<i1>.to<impl1>()        |
    );                                  |
                                        |
auto injector = di::make_injector(      |
    di::bind<i1>.named(my).to(module)   |
);                                      |
```

*

<a id="providers"></a>
> **[Providers](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/providers.html)** | [Examples](https://github.com/krzysztof-jusiak/di/blob/cpp14/example/custom_provider.cpp)
* [heap](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/providers/heap.html)
* [stack\_over\_heap (default)](http://krzysztof-jusiak.github.io/di/cpp14/boost/libs/di/doc/html/di/concepts/providers/stack_over_heap_default.html)
```cpp
Heap no throw provider                  | Test

CHANGELOG.md
## [1.0.0] - TBD
### Added
- Support for MSVC-2015+
- Support for GCC-5+
- Support for CMake 2.8+
- Support for libc++, libstdc++
- Support for clang-format

## [0.5.0] - 2015-01-12
### Added
- C++14 version
- Support for Clang3.4+

## [0.1.0] - 2014-08-15
### Added
- C++03/C++11 version
- Support for Clang3.2+

[1.0.0]: https://github.com/krzysztof-jusiak/di/compare/v0.5.0...v1.0.0
[0.5.0]: https://github.com/krzysztof-jusiak/di/compare/v0.1.0...v0.5.0
[0.1.0]: https://github.com/krzysztof-jusiak/di/tree/v0.1.0


tools/pph.sh
#!/bin/bash
#
# Copyright (c) 2012-2015 Krzysztof Jusiak (krzysztof at jusiak dot net)
#
# Distributed under the Boost Software License, Version 1.0.
# (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
#

pph() {
	version=$1 revision=$2 patch=$3
    echo "//"
    echo "// Copyright (c) 2012-2015 Krzysztof Jusiak (krzysztof at jusiak dot net)"
    echo "//"
    echo "// Distributed under the Boost Software License, Version 1.0."
    echo "// (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)"
    echo "//"
    echo "#pragma once"
    echo "#if (__cplusplus < 201305L && _MSC_VER < 1900)"
    echo "#error \"Boost.DI requires C++14 support (Clang-3.4+, GCC-5.1+, MSVC-2015+)\""
    echo "#else"
    echo "#define BOOST_DI_VERSION ${version}'${revision}'${patch}"
    echo "#define BOOST_DI_NAMESPACE_BEGIN namespace boost { namespace di { inline namespace v${version}_${revision}_${patch} {"
    echo "#define BOOST_DI_NAMESPACE_END }}}"
    echo "#if !defined(BOOST_DI_CFG_DIAGNOSTICS_LEVEL)"
    echo "#define BOOST_DI_CFG_DIAGNOSTICS_LEVEL 1"
    echo "#endif"
    echo "#if defined(BOOST_DI_CFG_FWD)"
    echo "BOOST_DI_CFG_FWD"
    echo "#endif"
    rm -rf tmp && mkdir tmp && cp -r boost tmp && cd tmp
    find . -iname "*.hpp" | xargs sed -i "s/BOOST_DI_NAMESPACE/::boost::di::v${version}_${revision}_${patch}/g"
    find . -iname "*.hpp" | xargs sed -i "s/\(.*\)__pph__/\/\/\/\/\1/g"
    find . -iname "*.hpp" | xargs sed -i "s/.*\(clang-format.*\)/\/\/\/\/\1/g"
    tail -n +10 "boost/di/aux_/compiler.hpp" | head -n -2 | sed '/^$/d' | sed "s/ \/\/\\(.*\)//g" | sed "s/\/\/\/\///"
    echo "#if defined(__CLANG__)"
    echo "#if (!BOOST_DI_CFG_DIAGNOSTICS_LEVEL)"
    echo "#pragma clang diagnostic error \"-Wdeprecated-declarations\""
    echo "#else"
    echo "#pragma clang diagnostic warning \"-Wdeprecated-declarations\""
    echo "#endif"
    echo "#pragma clang diagnostic push"
	echo "#pragma clang diagnostic error \"-Wundefined-inline\""
	echo "#pragma clang diagnostic error \"-Wundefined-internal\""
    echo "#elif defined(__GCC__)"
    echo "#pragma GCC diagnostic error \"-Wdeprecated-declarations\""
    echo "#pragma GCC diagnostic push"
	echo "#if (__GNUC__ < 6)"
	echo "#pragma GCC diagnostic error \"-Werror\""
	echo "#endif"
    echo "#elif defined(__MSVC__)"
	echo "#pragma warning(disable : 4503)"
	echo "#pragma warning(disable : 4822)"
	echo "#pragma warning(disable : 4505)"
    echo "#endif"
    echo '
        #include "boost/di/fwd_ext.hpp"
		BOOST_DI_NAMESPACE_BEGIN
        #include "boost/di/config.hpp"
        #include "boost/di/bindings.hpp"
        #include "boost/di/inject.hpp"
        #include "boost/di/injector.hpp"
        #include "boost/di/make_injector.hpp"
        #include "boost/di/scopes/deduce.hpp"
        #include "boost/di/scopes/instance.hpp"
        #include "boost/di/scopes/singleton.hpp"
        #include "boost/di/scopes/unique.hpp"
        #include "boost/di/policies/constructible.hpp"
        #include "boost/di/providers/heap.hpp"
        #include "boost/di/providers/stack_over_heap.hpp"' > tmp.hpp
    cpp -C -P -nostdinc -I. \
        -DBOOST_DI_AUX_COMPILER_HPP \
        -DBOOST_DI_AUX_PREPROCESSOR_HPP \
        -DBOOST_DI_INJECT_HPP tmp.hpp 2>/dev/null | \
            sed "s/\/\/\/\///" | \
            sed "s/[ $]*#define/##define/g" | \
            cpp -P -I. -fpreprocessed - 2>/dev/null | \
			sed "s/clang-format\(.*\)/\/\/ clang-format\1/g" | \
            sed "s/^##define/#define/g"
    tail -n +11 "boost/di/aux_/preprocessor.hpp" | head -n -2 | sed '/^$/d' | sed "s/ \/\/\\(.*\)//g"
    tail -n +15 "boost/di/inject.hpp" | head -n -2 | sed '/^$/d' | sed "s/ \/\/\\(.*\)//g"
    cd .. && rm -rf tmp
    echo "BOOST_DI_NAMESPACE_END"
    echo "#endif"
    echo "#if defined(__CLANG__)"
    echo "#pragma clang diagnostic pop"
    echo "#elif defined(__GCC__)"
    echo "#pragma GCC diagnostic pop"
    echo "#endif"
    echo
}

set -e
cpp --version >/dev/null
xargs --version >/dev/null
sed --version >/dev/null
head --version >/dev/null
tail --version >/dev/null

cd ${0%/*}/../include && pph `head -1 ../CHANGELOG.md  | sed "s/.*\[\(.*\)\].*/\1/" | tr '.' ' '` > "boost/di.hpp"
${CLANG_FORMAT:=clang-format} -i "boost/di.hpp"

